finetuner/tuner/dataset/__init__.py

```diff
@@ -1,8 +1,184 @@
+import abc
 import itertools
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 import numpy as np
 
 from ... import __default_tag_key__
+from ...helper import DocumentSequence
+
+
+class BaseDataset:
+    @abc.abstractmethod
+    def __init__(
+        self,
+        docs: DocumentSequence,
+        preprocess: Optional[Callable[[Union[str, np.ndarray]], Any]] = None,
+    ):
+        ...
+
+    @abc.abstractmethod
+    def __getitem__(
+        self, ind: int
+    ) -> Tuple[Union[str, np.ndarray], Union[int, Tuple[int, int]]]:
+        ...
+
+    @property
+    @abc.abstractmethod
+    def labels(self) -> Union[int, Tuple[int, int]]:
+        ...
+
+    @abc.abstractmethod
+    def __len__(self) -> int:
+        ...
+
+
+class ClassDataset:
+    """ Dataset for enapsulating data where each item has a class label."""
+
+    def __init__(
+        self,
+        docs: DocumentSequence,
+        preprocess: Optional[Callable[[Union[str, np.ndarray]], Any]] = None,
+    ):
+        """Create the :class:`ClassDataset` instance.
+
+        :param docs: The documents for the dataset. Each document is expected to have
+            - a content (only blob or text are accepted currently)
+            - a class label, saved under ``tags['finetuner']['label']``. This class
+              label should be an integer or a string
+        :param preprocess: A pre-processing function. It should take as input the
+            content of an item in the dataset (currently only text of blob are
+            accepted).
+        """
+        self._docs = docs
+        self._preprocess = preprocess
+
+        self._tag_labels_dict: Dict[Union[str, int], int] = {}
+        self._labels: List[int] = []
+        max_label = 0
+
+        for doc in self._docs:
+            try:
+                tag = doc.tags['finetuner']['label']
+            except KeyError:
+                raise KeyError(
+                    'The tag ["finetuner"]["label"] was not found in a document.'
+                    ' When using ClassDataset all documents need this tag'
+                )
+            label = self._tag_labels_dict.get(tag)
+
+            if label is None:
+                label = max_label
+                self._tag_labels_dict[tag] = label
+                max_label += 1
+
+            self._labels.append(label)
+
+    def __getitem__(self, ind: int) -> Tuple[Union[np.ndarray, str], int]:
+        """
+        Get the (preprocessed) content and label for the item at ``ind`` index in the
+        dataset.
+        """
+        content = self._docs[ind].content
+        label = self._labels[ind]
+
+        if self._preprocess is not None:
+            content = self._preprocess(content)
+
+        return (content, label)
+
+    @property
+    def labels(self) -> List[int]:
+        """ Get the list of integer labels for all items in the dataset."""
+        return self._labels
```

Comment:
this can be done in base class

Reply:
Removed the base class as it did not really much. Besides, I want this to be different as methods in different class have different docstrings